### ISSUE TITLE
Alien Ressurection doors won't open.

### ISSUE THREAD

**Issue by LaFlex** [MiSTer-devel/PSX_MiSTer#182]
Issue:
Certain doors (2 discovered so far) will not open after reloading savegame from savedata.

Reproduce:
Warehouse elevator door on level one (immediately after first save location) will only open on a new game. Once you die and reload your savegame, the door refuses to open. You will hear "Warehouse Access Granted" but the elevator door remains closed.

There is another door on mission three that also will not open. You will hear "security seal enabled" even after triggering the switch above said door to unlock it. Unsure if starting a new game and making it this far works.

Workaround:
Using savestates to avoid death works on level 1. Unsure if this carries over to the mission 3 door.

BIOS
1001

CD Image
bin/cue

Core Version / Build date
V220813

I have tested this with various CD options on and off. Nothing seems to affect this. For me it is only triggered when you actually load the save data from either a virtual or physical memory card.

**Comment by Kuba-J** [comment]
please share save state

**Comment by Fuzzybane** [comment]
fixed in 220815 build as far as I can tell, I saved the game on mission 1, died, reloaded mission 1 save, picked up the keycard again and opened the door just fine, I will try and get to mission 3 to confirm the other door

**Comment by Kuba-J** [comment]
https://www.reddit.com/r/AskProgramming/comments/7xux3y/why_does_the_infamous_door_glitch_in_alien/

**Comment by Fuzzybane** [comment]
that explains why it now works with trhe 32bit/1 cycle memory access, the core is replicating the original hardware almost 1:1 now

**Comment by LaFlex** [comment]
savestate after using the save portal
https://drive.google.com/file/d/1ka3eXlQAI51vKf0uQgYTfPAekLCJEwUW/view?usp=sharing

confirmed not working again just this morning.

**Comment by Fuzzybane** [comment]
> savestate after using the save portal https://drive.google.com/file/d/1_7AzXgF0EaBJRh4LOWc5v4wTD5V3wFDN/view?usp=sharing
> 
> confirmed not working again just this morning.

have you tried it with the latest unstable or just the latest stable release?

**Comment by LaFlex** [comment]
I have not tried it with the latest unstable release. Unfortunately I won't be able to get in front of my mister for another 12 hours today. I will report back tonight

**Comment by Fuzzybane** [comment]
no worries, either @ me here or in the psx-chat channel on the discord, my user name there is just Fuzz

**Comment by Kuba-J** [comment]
Also happens on the real PSX, strange bug

https://www.youtube.com/watch?v=bXIW2WGZnhQ

**Comment by Fuzzybane** [comment]
the way I did it yesterday was I saved before getting the keycard, reloaded the save after death went back down to where the gun and keycard was and then opened the door and it was fine, maybe I got lucky?

**Comment by LaFlex** [comment]
> 

You got it! I'm in! As long as you get the key in the same run that you go for the door it seems to work.

Getting the keycard _after_ using the save terminal seems to be the trigger for me.
I did it your way, first saving in-game before getting the keycard, then I got the card, progressed through the door, died, reloaded the save in-game. Can anyone confirm this is how it works on a stock ps1? If so i think we can close this post.

Now please, lets get me in door #2. It's on mission 3. The door that says "Security Seal Enabled" when you approach it.

**Comment by LaFlex** [comment]
I brought a new savefile i created on mission 3 right before the door that won't open. According to my sources you should be able to walk up the ladder, press the button (no need to get the card on the floor yet), and open the door below.

I cannot get in this door, i have tried everything, up to and including starting the entire mission over from scratch and not dying/reloading at all.

https://drive.google.com/file/d/1wihtPBdS3xeXmWdkmzMDW-bd5V5cm7lh/view?usp=sharing

Something is definitely wrong and its the same something that was wrong on mission 1.

I was able to get my self locked out of an earlier door on mission three but got around it by very specifically pressing the button for the door, and then immediately going to the door to confirm it opens, then going back to where the switch was to gather more items and an ID tag.

**Comment by Fuzzybane** [comment]
That is unfortunate, hold onto this save file as it will be a good test fort when all modules timings are accurate and see ifd this gets fully fixed

**Comment by RobertPeip** [comment]
[Alien Resurrection (USA)_2.zip](https://github.com/MiSTer-devel/PSX_MiSTer/files/9438246/Alien.Resurrection.USA._2.zip)

Adding a savestate here that can be used for fast reproducing (thanks to Fuzz for creating it)

Using this savestate it can be seen that the door currently does not open.
If you however let the sound play and then disable sound in the debug menu for a short moment, the door opens.

**Comment by Natrox** [comment]
Tested and reproduced on latest unstable (as of writing), PSX_unstable_20220903_06b0b5.rbf, all options stock (no unsafe (U) settings used). Used Robert's SS (thanks) and also tried the other ones posted. 

Redump version (redumpSonyPlaystationAmericaGames02October2017), multitrack. MD5 of track 1 (~257MB): 0542c7e2ed28db4aaffc1534e8760d78

If you would like me to test again, I propose it best if you give me a memory card image so I can be more sure. It's probably easier to just grab the latest unstable yourself for this.

**Comment by RobertPeip** [comment]
Analyzed what happens at the door opening:

CD streams data to ram before SPU DMA, not timing critical
Data is transfered with SPU DMA before it's read by SPU, not timing critical
data is 0x010C at 0x4fe(=0x27F0) => contains loop flag, leading to endless loop from 0x4fe to 0x202
SPU IRQ waits for voice to reach 0x502 to trigger IRQ, which never happens due to the loop.

Any hack to continue voice address increase will open the door

**Comment by RobertPeip** [comment]
Workaround now exists:
- activate debug menu with L3+R3+Select while OSD is closed
- in Debug menu select "Sound IRQ Trigger" after the voice is finished
-> door will open

**Comment by a-barone** [comment]
I discovered that there are two prototype builds available on Hidden Palace that allows the level 1 door to be opened at all times, although with a caveat (see the recording below).
The prototype builds that work are listed below, and they are the first builds available there with a saving system implemented:
https://hiddenpalace.org/Alien_Resurrection_(May_26,_2000_prototype)
https://hiddenpalace.org/Alien_Resurrection_(May_30,_2000_prototype)

Test scenario recording for the May 30, 2000 build:
https://youtu.be/yrNNxtIMzRE

And here's the same test scenario but using the official US release (by the way, the PAL release behaves the same way; I've tested it too):
https://youtu.be/zQy1CfdcyOc

One important detail that may have gone unnoticed is:
- The official releases DO seem to have a software bug in them to the extent that if you collect the access key during the gameplay and try to open the door, it does NOT play the "Warehouse access granted" voice sample.
But if you load a game with the access key already saved to the inventory, the game ALWAYS plays the "Warehouse access granted" voice sample.
So, in fact, some part of the game logic is only executed if you load the game with the access key already in the inventory; and that's not a fault of the PSX Core.

- In the late May 2000 prototypes, that voice sample is also only played IF you load the game with the access card already in the inventory, but, *unlike the later/final versions*, it does allow you to: 1) open the door with only one try once your presence is detected (see the first video linked in this post); 2) open the door in a second activation of the switch (which is not possible in the later/final builds), which displays the text message again but does NOT play that "Warehouse access granted" voice sample, thus immediately opening the door.
